crates/miden-objects/src/account/component/metadata/storage/toml/tests.rs

```diff
@@ -0,0 +1,714 @@
+use crate::errors::AccountComponentTemplateError;
+
+#[test]
+fn from_toml_str_with_nested_table_and_flattened() {
+    let toml_table = r#"
+        [token_metadata]
+        max_supply = "1000000000"
+        symbol = "ETH"
+        decimals = "9"
+    "#;
+
+    let toml_inline = r#"
+        token_metadata.max_supply = "1000000000"
+        token_metadata.symbol = "ETH"
+        token_metadata.decimals = "9"
+    "#;
+
+    let storage_table = InitStorageData::from_toml(toml_table).unwrap();
+    let storage_inline = InitStorageData::from_toml(toml_inline).unwrap();
+
+    assert_eq!(storage_table.values(), storage_inline.values());
+}
+
+#[test]
+fn from_toml_str_with_deeply_nested_tables() {
+    let toml_str = r#"
+        [a]
+        b = "0xb"
+
+        [a.c]
+        d = "0xd"
+
+        [x.y.z]
+        w = "42"
+    "#;
+
+    let storage = InitStorageData::from_toml(toml_str).expect("Failed to parse TOML");
+    let key1: StorageValueName = "a.b".parse().unwrap();
+    let key2: StorageValueName = "a.c.d".parse().unwrap();
+    let key3: StorageValueName = "x.y.z.w".parse().unwrap();
+
+    let WordValue::Scalar(v1) = storage.get(&key1).unwrap() else {
+        panic!("expected a raw value for {key1}");
+    };
+    let WordValue::Scalar(v2) = storage.get(&key2).unwrap() else {
+        panic!("expected a raw value for {key2}");
+    };
+    let WordValue::Scalar(v3) = storage.get(&key3).unwrap() else {
+        panic!("expected a raw value for {key3}");
+    };
+
+    assert_eq!(v1, "0xb");
+    assert_eq!(v2, "0xd");
+    assert_eq!(v3, "42");
+}
+
+#[test]
+fn from_toml_rejects_non_string_scalars() {
+    let toml_str = r#"
+        foo = 42
+    "#;
+
+    let result = InitStorageData::from_toml(toml_str);
+    assert_matches::assert_matches!(result.unwrap_err(), InitStorageDataError::NonStringScalar(_));
+}
+
+#[test]
+fn test_error_on_array() {
+    let toml_str = r#"
+        token_metadata.v = [1, 2, 3]
+    "#;
+
+    let result = InitStorageData::from_toml(toml_str);
+    assert_matches::assert_matches!(result.unwrap_err(), InitStorageDataError::ArraysNotSupported);
+}
+
+#[test]
+fn parse_map_entries_from_array() {
+    let toml_str = r#"
+        my_map = [
+            { key = "0x0000000000000000000000000000000000000000000000000000000000000001", value = "0x0000000000000000000000000000000000000000000000000000000000000010" },
+            { key = "0x0000000000000000000000000000000000000000000000000000000000000002", value = ["1", "2", "3", "4"] }
+        ]
+    "#;
+
+    let storage = InitStorageData::from_toml(toml_str).expect("Failed to parse map entries");
+    let map_name: StorageValueName = "my_map".parse().unwrap();
+    let entries = storage.map_entries(&map_name).expect("map entries missing");
+    assert_eq!(entries.len(), 2);
+
+    assert_matches::assert_matches!(
+        &entries[0].0,
+        WordValue::Scalar(v)
+            if v == "0x0000000000000000000000000000000000000000000000000000000000000001"
```

Comment:
```suggestion
        WordValue::Scalar(k)
            if k == "0x0000000000000000000000000000000000000000000000000000000000000001"
```

to distinguish the key from the value in the map entry